### PR TITLE
Improve get partitions and reset spec

### PR DESCRIPTION
**Pull request by Itxaka**
- do not fill mountpoints if we dont need them
 - try to unmount partitions before we format them
 - only warn about missing partitions if we are asking to format them
 - getpartitions now tries to find the mountpoint by label as well

The main issue here was that a partition mounted as /dev/disk/by-label/LABEL would not be found properly by ghw library, so the mountpoint will always be empty

This patch introduces an extra check on the GetAllPartitions functions that if the mountpoint its empty it will try to infer if it was mounted by label instead and gather the mountpoint

Only reset should be affected by this, as we only format in that case and we only need the mountpoint during the reset to be able to unmount it beforehand. In upgrade and install we dont really need that so its not implemented there.

Fixes: https://github.com/kairos-io/kairos/issues/2836

### CODE REVIEW

File: pkg/config/spec.go [hunk 1]
```diff
@@ -438,34 +438,18 @@ func NewResetSpec(cfg *Config) (*v1.ResetSpec, error) {
...
 		// We could have oem in lvm which won't appear in ghw list
 		ep.OEM = partitions.GetPartitionViaDM(cfg.Fs, constants.OEMLabel)
 	}
 
-	if ep.OEM == nil {
-		cfg.Logger.Warnf("no OEM partition found")
-	}
-
 	// Persistent partition is not a hard requirement
-	if ep.Persistent != nil {
```
Comment (Itxaka) [inline pkg/config/spec.go:457]: removed clutter here and above, not really needed.

File: pkg/config/spec.go [hunk 2]
```diff
@@ -522,6 +506,25 @@ func NewResetSpec(cfg *Config) (*v1.ResetSpec, error) {
 		return nil, fmt.Errorf("failed unmarshalling the full spec: %w", err)
 	}
 
+	if ep.OEM == nil && spec.FormatOEM {
+		cfg.Logger.Warnf("no OEM partition found, wont format it")
```
Comment (Itxaka) [inline pkg/config/spec.go:510]: only warn if we are actually formatting the partition

File: pkg/config/spec.go [hunk 3]
```diff
@@ -522,6 +506,25 @@ func NewResetSpec(cfg *Config) (*v1.ResetSpec, error) {
...
+	}
+
+	if ep.Persistent == nil && spec.FormatPersistent {
+		cfg.Logger.Warnf("no Persistent partition found, wont formate it")
+	}
+
+	// If we mount partitions by the /dev/disk/by-label stanza, their mountpoints wont show up due to ghw not
```
Comment (Itxaka) [inline pkg/config/spec.go:517]: this is the real change that fixes this + the unmount

File: pkg/elemental/elemental.go [hunk 4]
```diff
@@ -47,14 +47,7 @@ func NewElemental(config *agentConfig.Config) *Elemental {
 
 // FormatPartition will format an already existing partition
 func (e *Elemental) FormatPartition(part *v1.Partition, opts ...string) error {
-	var name string
-	// Nice display name for logs
```
Comment (Itxaka) [inline pkg/elemental/elemental.go:51]: not really neede, we can use the label directly?

File: pkg/utils/partitions/getpartitions.go [hunk 5]
```diff
@@ -61,6 +64,59 @@ func GetAllPartitions() (v1.PartitionList, error) {
 	return parts, nil
 }
 
+// GetMountPointByLabel will try to get the mountpoint by using the label only
+// so we can identify mounts the have been mounted with /dev/disk/by-label stanzas
+func GetMountPointByLabel(label string) string {
```
Comment (Itxaka) [inline pkg/utils/partitions/getpartitions.go:69]: at this point, reimplementing what GHW does its just a few lines more and we can adapt to our own specifics...

File: pkg/config/spec.go [hunk 6]
```diff
@@ -522,6 +506,25 @@ func NewResetSpec(cfg *Config) (*v1.ResetSpec, error) {
 		return nil, fmt.Errorf("failed unmarshalling the full spec: %w", err)
 	}
 
+	if ep.OEM == nil && spec.FormatOEM {
+		cfg.Logger.Warnf("no OEM partition found, wont format it")
+	}
+
+	if ep.Persistent == nil && spec.FormatPersistent {
+		cfg.Logger.Warnf("no Persistent partition found, wont formate it")
```
Comment (jimmykarily) [inline pkg/config/spec.go:514]:
```suggestion
		cfg.Logger.Warnf("no Persistent partition found, won't format it")
```

File: pkg/config/spec.go (hunk 2 above)
Comment (jimmykarily) [inline pkg/config/spec.go:510]:
```suggestion
		cfg.Logger.Warnf("no OEM partition found, won't format it")
```

File: pkg/config/spec.go (hunk 1 above)
Comment (jimmykarily) [reply pkg/config/spec.go:457]: It's not clear why it's not needed but I trust you :). It seems to be setting the default for some reason.

**Review by jimmykarily — CHANGES_REQUESTED** [review]
Since this is the agent, this may not be covered by any tests here but we should probably add a test in the e2e suite to make sure we don't break it again.

**Comment by Itxaka** [comment]
> Since this is the agent, this may not be covered by any tests here but we should probably add a test in the e2e suite to make sure we don't break it again.

We have a reset test already in the e2e suite of kairos. We dont have a test for reset-oem as that would need extra work to support it, but we should do one. Something like reset-oem + after-reset-chroot stage to create a new cloud-config so on restart it can login :D

Also there is unit tests for this I think which should cover this changes.

**Comment by jimmykarily** [comment]
> > Since this is the agent, this may not be covered by any tests here but we should probably add a test in the e2e suite to make sure we don't break it again.
> 
> We have a reset test already in the e2e suite of kairos. We dont have a test for reset-oem as that would need extra work to support it, but we should do one. Something like reset-oem + after-reset-chroot stage to create a new cloud-config so on restart it can login :D
> 
> Also there is unit tests for this I think which should cover this changes.

If it was covered, shouldn't those tests be broken?

File: pkg/action/reset.go [hunk 7]
```diff
@@ -114,6 +114,7 @@ func (r ResetAction) Run() (err error) {
 	defer func() { err = cleanup.Cleanup(err) }()
 
 	// Unmount partitions if any is already mounted before formatting
+	// TODO: Is this needed???
```
Comment (jimmykarily) [inline pkg/action/reset.go:117]: Who can answer that? :D

**Review by jimmykarily — APPROVED** [review]

**Comment by Itxaka** [comment]
> > > Since this is the agent, this may not be covered by any tests here but we should probably add a test in the e2e suite to make sure we don't break it again.
> > 
> > We have a reset test already in the e2e suite of kairos. We dont have a test for reset-oem as that would need extra work to support it, but we should do one. Something like reset-oem + after-reset-chroot stage to create a new cloud-config so on restart it can login :D
> > 
> > Also there is unit tests for this I think which should cover this changes.
> 
> If it was covered, shouldn't those tests be broken?

Only the normal reset ones, we do not test the reset OEM part :)

That will change hopefully and we will see this patch fixing the test I hope?

File: pkg/action/reset.go (hunk 7 above)
Comment (Itxaka) [reply pkg/action/reset.go:117]: tests probably!

**Pull request merged**